File: solutions/ruby/isbn-verifier/4/isbn_verifier.rb
class IsbnVerifier

  VALID_ISBN10_FORMAT = /\A[[:digit:]]{9}[[:digit:]X]\z/

  def self.valid? isbn
    new(isbn).validate
  end

  private

  attr_accessor :isbn

  def initialize isbn
    self.isbn = isbn.delete '-'
  end

  def digit_value digit
    digit == 'X' ? 10 : digit.to_i
  end

  def checksum
    isbn.each_char.with_index.sum { |digit, i|
      digit_value(digit) * (10 - i)
    }
  end

  public

  def validate
    isbn.match?(VALID_ISBN10_FORMAT) &&
      checksum.modulo(11).zero?
  end

end

File: solutions/elixir/isbn-verifier/4/lib/isbn_verifier.ex
defmodule IsbnVerifier do

  def isbn?(isbn), do: valid?(isbn, 10, 0)

  defp valid?("", 0, checksum), do: rem(checksum, 11) == 0

  defp valid?(<< ?- , rest::binary >>, position, checksum),
    do: valid?(rest, position, checksum)

  defp valid?(<< ?X , rest::binary >>, 1, checksum),
    do: valid?(rest, 0, checksum + 10)

  defp valid?(<< digit, rest::binary >>, position, checksum)
       when digit in ?0..?9 and position > 0 do
    valid?(rest, position - 1, checksum + (digit - ?0) * position)
  end

  defp valid?(_, _, _), do: false

end
